### PR TITLE
MAINT: Clean up all module-level dunders

### PR DESCRIPTION
**Pull request by BvB93**
This PR attempts the clean up the module-level dunders by means of the following two actions:
* Replacing `__all__` with explicit reexports. The use of `__all__` is a bit of remnant, one from before I 
 was aware of `from X import Y as Y` . Note that both syntaxes are formally equivalent.
* Add annotations for a number of, previously missing, dunders such as `__version__`.

### CODE REVIEW

File: numpy/__init__.pyi [hunk 1]
```diff
@@ -311,52 +311,13 @@ from numpy.core.shape_base import (
...
-    "size",
-    "around",
-    "mean",
-    "std",
-    "var",
-]
+__all__: List[str]
+__path__: List[str]
+__version__: str
+__git_version__: str
```
Comment (BvB93) [inline numpy/__init__.pyi:317]: I'm not 100% sure if all these dunders are considered public or not (especially the bottom three). Are there any about annotating them?

Secondly, they could be marked with `Final` if they should be considered read-only.

Comment (eric-wieser) [reply numpy/__init__.pyi:317]: `__path__` is a python builtin that's added to all packages, like `__name__`.

Comment (BvB93) [reply numpy/__init__.pyi:317]: The annoying thing it is not defined for [`types.ModuleType`](https://github.com/python/typeshed/blob/a1a51f5aace0989301576c7f7edc7432bef6f5c9/stdlib/3/_importlib_modulespec.pyi#L35-L42) (as not all modules are packages), and by extension for `numpy`.
This is arguably an issue that should be solved in mypy though (or perhaps even typeshed).

Comment (BvB93) [reply numpy/__init__.pyi:317]: How about the likes of `np.__NUMPY_SETUP__`, `np.__deprecated_attrs__` and `np.__expired_functions__` though?
They're not (clearly) marked as private, but I'm not 100% sure whether or they're considered public.

Comment (seberg) [reply numpy/__init__.pyi:317]: I would consider them private, maybe they should never have been a dunder as such.  I am least sure about `__NUMPY_SETUP__`, but frankly, I wonder if that will matter in a typing context.

Comment (BvB93) [reply numpy/__init__.pyi:317]: Yeah, for now I've just removed them (https://github.com/numpy/numpy/pull/18229/commits/7060d5e28f735199df6b465755ddd1d3bc7ed07e).
We can always consider revisiting this subject if the need arises.

**Review by rgommers — APPROVED** [review]
LGTM, thanks @BvB93

**Pull request merged**